Fix xml nesting bug in CustomSerializer.WriteMemberInfoCollection()

IISResetMe opened the pull request:
## PR Summary
Fix #8308 

`WriteMemberInfoCollection()` incorrectly nests sibling elements at depth==0
The reason seems to be that `WriteMemberInfoCollection()` calls `WriteEndElement()` at most once, even if `WriteStartElement()` has been called more than once.

This PR moves the `WriteEndElement()` call up immediately after the elements text value has been written.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

IISResetMe committed:
Remove incorrect modification of `depth` parameter

Decrementing `depth` makes no sense here and causes nested PSObject instances with more than one property to generate invalid xml

IISResetMe committed:
Revert "Remove incorrect modification of `depth` parameter"

This reverts commit 56e07a011858c6dd8c8d7373e2fb7d0aadd662e1.

IISResetMe committed:
Fix #8308

WriteMemberInfoCollection() incorrectly nests sibling elements at depth==0
The reason seems to be that WriteMemberInfoCollection() calls WriteEndElement() at most once, even if `WriteStartElement()` has been called more than once

IISResetMe committed:
Add tests for with ConvertTo-Xml with nested PSObjects

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Xml.Tests.ps1:
```diff
@@ -108,5 +108,20 @@ Describe "ConvertTo-Xml DRT Unit Tests" -Tags "CI" {
         $x.Objects.Object[1].Property."#text" | Should -BeExactly "banana"
     }
+
+    It "Serialize nested PSObject properties" {
```
```suggestion
    It "Serialize nested pscustomobject properties" {
```

iSazonov committed:
Update test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Xml.Tests.ps1

Co-Authored-By: IISResetMe <IISResetMe@users.noreply.github.com>

PaulHigin commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CustomSerialization.cs:
```diff
@@ -691,7 +691,6 @@ private bool PSObjectHasNotes(PSObject source)
         {
```
These changes look right to me ... not sure what was originally intended.  But can you also remove the `depth` and `writeEnclosingememberSetElementTag` parameters?  They are not used within the method and it would help clean up the code by removing them (also please remove the xml comments for the parameters).

IISResetMe replied:
Done, I renamed the first parameter to `members` as well

IISResetMe committed:
Clean up WriteMemberInfoCollection()

Removed unused parameters `depth` and `writeEnclosingMemberSetElementTag`, renamed `me` to `members`

IISResetMe committed:
Merge branch 'patch/invalid-xml-CustomSerialization' of https://github.com/IISResetMe/PowerShell into patch/invalid-xml-CustomSerialization

PaulHigin reviewed (approved):
LGTM

iSazonov commented:
The original code is obviously broken.
As for the depth, I am afraid that this fix is not entirely correct. I assume that recursion was originally meant to serialize members.
@mklement0  What do you think?

IISResetMe commented:
@iSazonov I agree that depth limitations are not at all correctly imposed by the serializer right now, but that's a broader issue that should be addressed in a separate issue/PR. This is simply an attempt to not produce invalid XML by fixing an adjecent bug.

For the broader issue of depth limitations we'll need to address the fact that depth is currently ignored by container types as well.

IISResetMe commented:
I'll revert the last commit, but plan on tackle the depth issue as part of https://github.com/PowerShell/PowerShell/issues/7290

IISResetMe committed:
Revert WriteMemberInfoCollection cleanup commit

IISResetMe commented:
@iSazonov Reverted `depth`-related refactorings

iSazonov commented:
@IISResetMe Please add a empty commit with `[Feature]` in title before we merge..

IISResetMe committed:
[Feature]Trigger tests

iRon7 commented:
Please be aware that this issue shouldn't only concern the _technical_ fix of #7290 but also the presumed compatibility with prior versions. Meaning: a lot of scripts in the field might break due to this change!
Refer to the discussion in case #8393 from @mklement0  concerning the perception of the default  `-Depth` (of **`2`**) of the similar cmdlet `ConvertTo-Json`.

iSazonov commented:
@iRon7 In the PR we fix a bug in ConvertTo-Xml cmdlet but you reference a ConvertTo-Json - could you clarify that you mean?

iRon7 commented:
I am sorry, I guess my comment wasn't that clear. What I mean is, that appearently every user of the `ConvertTo-Json` cmdlets appears to trapped by the pitfall that the `-Depth`is limited by default to `2`. The issues with `ConvertTo-Xml` are less than `ConvertTo-Json` which could be explained by usage but also by the concerned bug. Meaning that if the default depth now suddenly changes to `2`, users might experience the same type of issues as with `ConvertTo-Json`. Besides, scripts that use to work might now suddenly break and have truncated property paths...

IISResetMe commented:
@iRon7 one approach could be:

 - Fix depth enforcement 
 - Set default depth = 5

Doesnt have to be 2 just because that's the case for the json cmdlet

iRon7 commented:
Yes, that might work but could cause an performance impact for situation were the default depth did work.
E.g.:
```PowerShell
Get-Service | ConvertTo-Xml
```
takes ~700ms, where:
```PowerShell
Get-Service | ConvertTo-Xml -Depth 5
```
takes 6.5 minutes...

PaulHigin commented:
@IISResetMe 
I still feel the unused method parameters should be removed.  Yes, they should eventually be implemented, but since they are not now, they should be removed.  It is Ok to add a TODO comment and create an Issue to add the functionality later.

iSazonov commented:
@PaulHigin A code you reference is not _new code_ so the change is not related the PR. I believe we get better code with fixing #7290.  @IISResetMe Are ready to fix #7290?

> takes 6.5 minutes...

@iRon7 Please open new performance issue so we can investigate this problem.

iRon7 commented:
@iSazonov, I do not have a performance issue (...yet). Just some concerns that the fix (even it is technically correct) might break a few scripts in the field and if you try to prevent that with a higher default `-Depth`, you _might_ introduce a few performance complains with existing scripts.
Anyway, it is totally up to you what to decide (any direction in this delimema has its pro's and cons).

iSazonov commented:
@SteveL-MSFT We need an additional review from MSFT experts and get a conclusion about @iRon7's concerns - can we merge before PowerShell Committee review #7290?

SteveL-MSFT commented:
I think the scope of these changes is ok to accept without waiting on conclusion of #7290

iRon7 commented:
I just found a similar issue with a `PSCustomObject` and a recurring reference:

```PowerShell
$Parent = [PSCustomObject]@{
	Name = "Parent"
}
$Child = [PSCustomObject]@{
	Name = "Child"
}
$Parent | Add-Member Child $Child
$Child | Add-Member Parent $Parent
```
But the error message is completely different from #7290

```PowerShell
PS C:\> $Parent | ConvertTo-Xml
```

> ConvertTo-Xml : Unexpected end of file has occurred. The following elements are not closed: Objects. Line 9, position 12.
> At line:1 char:11
> + $Parent | ConvertTo-Xml
> +           ~~~~~~~~~~~~~
>     + CategoryInfo          : NotSpecified: (:) [ConvertTo-Xml], XmlException
>     + FullyQualifiedErrorId : System.Xml.XmlException,Microsoft.PowerShell.Commands.ConvertToXmlCommand 

I am not sure whether this is covered in the fix and tests
Anyways, let me know if you like me to open a new bug report for this.

iSazonov commented:
@iRon7 You can download a latest night build from main page of the repo and test.
For your example I get:
```
$Parent | ConvertTo-Xml

xml                            Objects
---                            -------
version="1.0" encoding="utf-8" Objects
```
